src/lib/Lower.hs

```diff
@@ -365,58 +403,71 @@ fromNameVAtom v = case eqColorRep @c @AtomNameC of
   _ -> error "Unexpected non-atom name"
 
 newtype VectorizeM i o a =
-  VectorizeM { runVectorizeM :: SubstReaderT VSubstValC (BuilderT SimpIR (ReaderT Word32 HardFailM)) i o a }
+  VectorizeM { runVectorizeM :: SubstReaderT VSubstValC (BuilderT SimpIR (ReaderT Word32 FallibleM)) i o a }
   deriving ( Functor, Applicative, Monad, Fallible, MonadFail, SubstReader VSubstValC
            , Builder SimpIR, EnvReader, EnvExtender, ScopeReader, ScopableBuilder SimpIR)
 
 getLoopWidth :: VectorizeM i o Word32
 getLoopWidth = VectorizeM $ SubstReaderT $ ReaderT $ const $ ask
 
 vectorizeBlock :: Emits o => SBlock i -> VectorizeM i o (VAtom o)
-vectorizeBlock (Block _ decls (ans :: SAtom i')) = go decls
-  where
-    go :: Emits o => Nest SDecl i i' -> VectorizeM i o (VAtom o)
-    go = \case
-      Empty -> vectorizeAtom ans
-      Nest (Let b (DeclBinding ann _ expr)) rest -> do
-        unless (ann == PlainLet) $ error "Encountered a non-plain let?"
-        v <- vectorizeExpr expr
-        extendSubst (b @> v) $ go rest
+vectorizeBlock block@(Block _ decls (ans :: SAtom i')) =
+  addVectErrCtx "vectorizeBlock" ("Block:\n" ++ pprint block) $
+    go decls
+    where
+      go :: Emits o => Nest SDecl i i' -> VectorizeM i o (VAtom o)
+      go = \case
+        Empty -> vectorizeAtom ans
+        Nest (Let b (DeclBinding _ _ expr)) rest -> do
+          v <- vectorizeExpr expr
+          extendSubst (b @> v) $ go rest
 
 vectorizeExpr :: Emits o => SExpr i -> VectorizeM i o (VAtom o)
-vectorizeExpr expr = case expr of
-  Atom atom -> vectorizeAtom atom
-  Op op -> vectorizeOp op
-  -- Vectorizing IO might not always be safe! Here, we depend on vectorizeOp
-  -- being picky about the IO-inducing ops it supports, and expect it to
-  -- complain about FFI calls and the like.
-  Hof (RunIO body) -> do
+vectorizeExpr expr = addVectErrCtx "vectorizeExpr" ("Expr:\n" ++ pprint expr) do
+  case expr of
+    Atom atom -> vectorizeAtom atom
+    Op op -> vectorizeOp op
+    -- Vectorizing IO might not always be safe! Here, we depend on vectorizeOp
+    -- being picky about the IO-inducing ops it supports, and expect it to
+    -- complain about FFI calls and the like.
+    Hof (RunIO body) -> do
     -- TODO: buildBlockAux?
-    Abs decls (LiftE vy `PairE` yWithTy) <- buildScoped do
-      VVal vy y <- vectorizeBlock body
-      PairE (LiftE vy) <$> withType y
-    body' <- absToBlock =<< computeAbsEffects (Abs decls yWithTy)
-    VVal vy . Var <$> emit (Hof (RunIO body'))
-  _ -> error $ "Cannot vectorize expr: " ++ pprint expr
+      Abs decls (LiftE vy `PairE` yWithTy) <- buildScoped do
+        VVal vy y <- vectorizeBlock body
+        PairE (LiftE vy) <$> withType y
+      body' <- absToBlock =<< computeAbsEffects (Abs decls yWithTy)
+      VVal vy . Var <$> emit (Hof (RunIO body'))
+    _ -> throwVectErr $ "Cannot vectorize expr: " ++ pprint expr
+    -- Hof (RunIO (Lam (LamExpr (LamBinder b UnitTy PlainArrow _) body))) -> do
```

Comment:
This is presumably left over from bringing things up-to-date with main. Let's not bother with a roundtrip review cycle. I'll merge it as-is and delete it when I rebase my change on top of it, since that will require fixing merge conflicts in this file.